README.rst
csv2es
=========================

.. image:: https://img.shields.io/pypi/v/csv2es.svg
    :target: https://pypi.python.org/pypi/csv2es

.. image:: https://img.shields.io/travis/rholder/csv2es.svg
    :target: https://travis-ci.org/rholder/csv2es

.. image:: https://img.shields.io/pypi/dm/csv2es.svg
    :target: https://pypi.python.org/pypi/csv2es

The csv2es project is an Apache 2.0 licensed commandline utility, written in
Python, to load a CSV (or TSV) file into an Elasticsearch instance. That's
pretty much it. That's all it does. The first row of the file should contain
the field names intended to be used for Elasticsearch documents otherwise things
will get weird. There's a little trick documented below to add a header row in
case the file is missing it.


Features
--------

- Minimal commandline interface
- Load CSV's or TSV's
- Customize the delimiter to something else
- Uses the Elasticsearch bulk API
- Parallel bulk uploads
- Retry on errors with exponential backoff


Installation
------------

To install csv2es, simply:

.. code-block:: bash

    $ pip install csv2es


Usage
-----
::

 Usage: csv2es [OPTIONS]

   Bulk import a delimited file into a target Elasticsearch instance. Common
   delimited files include things like CSV and TSV.

   Load a CSV file:
     csv2es --index-name potatoes --doc-type potato --import-file potatoes.csv

   For a TSV file, note the tab delimiter option
     csv2es --index-name tomatoes --doc-type tomato --import-file tomatoes.tsv --tab

   For a nifty pipe-delimited file (delimiters must be one character):
     csv2es --index-name pipes --doc-type pipe --import-file pipes.psv --delimiter '|'

 Options:
   --index-name TEXT          Index name to load data into           [required]
   --doc-type TEXT            The document type (like user_records)  [required]
   --import-file TEXT         File with content to import            [required]
   --mapping-file TEXT        JSON mapping file for index
   --delimiter TEXT           The field delimiter to use, defaults to CSV
   --tab                      Assume tab-separated, overrides delimiter
   --host TEXT                The Elasticsearch host (http://127.0.0.1:9200/)
   --docs-per-chunk INTEGER   The documents per chunk to upload (5000)
   --bytes-per-chunk INTEGER  The bytes per chunk to upload (100000)
   --parallel INTEGER         Parallel uploads to send at once, defaults to 1
   --delete-index             Delete existing index if it exists
   --quiet                    Minimize console output
   --version                  Show the version and exit.
   --help                     Show this message and exit.


Examples
--------

Let's say we've got a potatoes.csv file with a nice header that looks like this::

 potato_id,potato_type,description
 33,sweet,"kinda oval"
 17,regular,bumpy
 91,regular,"perfectly round"
 18,sweet,delightful
 42,fried,crispy
 37,"extra special",crispy

Now we can stuff it into Elasticsearch:

.. code-block:: bash

    csv2es --index-name potatoes --doc-type potato --import-file potatoes.csv

But what if it was tomatoes.tsv and separated by tabs? Well, we can do this:

.. code-block:: bash

    csv2es --index-name tomatoes --doc-type tomato --import-file tomatoes.tsv --tab


Advanced Examples
-----------------

What if we have a super cool pipe-delimited file and want to wipe out the
existing "pipes" index every time we load it up? This ought to handle that case:

.. code-block:: bash

    csv2es --index-name pipes --delete-index --doc-type pipe --import-file pipes.psv --delimiter '|'

Elasticsearch is great, but it's doing something strange to our documents when
we try to facet by certain fields. Let's create our own custom mapping file to
specify the fields used in Elasticsearch for that potatoes.csv called
potatoes.mapping.json:

.. code-block:: json

    {
        "dynamic": "true",
        "properties": {
            "potato_id": {"type": "long"},
            "potato_type": {"type": "string", "index" : "not_analyzed"},
            "description": {"type": "string", "index" : "not_analyzed"},
        }
    }

Now let's load the data with a custom mapping file:

.. code-block:: bash

    csv2es --index-name potatoes --doc-type potato --mapping-file potatoes.mapping.json --import-file potatoes.csv

What if my file is missing the header row, and it's super huge because there are
so many potatoes in it, and everything is terrible? We can use sed to tack on a
nice header with something like this:

.. code-block:: bash

    sed -i 1i"potato_id,potato_type,description" potatoes.csv

As long as you have more disk space than the size of the file, this should be fine.


Contribute
----------

#. Check for open issues or open a fresh issue to start a discussion around a feature idea or a bug.
#. Fork `the repository`_ on GitHub to start making your changes to the **master** branch (or branch off of it).
#. Write a test which shows that the bug was fixed or that the feature works as expected.
#. Send a pull request and bug the maintainer until it gets merged and published. :) Make sure to add yourself to AUTHORS_.

.. _`the repository`: https://github.com/rholder/csv2es
.. _AUTHORS: https://github.com/rholder/csv2es/blob/master/AUTHORS.rst


HISTORY.rst
.. :changelog:

History
-------

1.0.2 (2016-05-31)
++++++++++++++++++
- Add option --settings to pass settings when creating new index (e.g. custom analyzers)

1.0.1 (2015-06-02)
++++++++++++++++++
- Add option to stream from stdin

1.0.0 (2015-04-23)
++++++++++++++++++
- Add retrying support with exponential backoff per chunk for bulk uploads
- Add parallel bulk uploading via joblib
- Stable release

1.0.0.dev3 (2015-04-19)
+++++++++++++++++++++++
- Switch over to Click for handling executable
- Fix --delete-index flag
- Add --version option

1.0.0.dev2 (2015-04-19)
+++++++++++++++++++++++
- Fix import errors

1.0.0.dev1 (2015-04-18)
+++++++++++++++++++++++
- Tinkering with documentation and PyPI updates

1.0.0.dev0 (2015-04-18)
+++++++++++++++++++++++
- First dev version now exists
- Apache 2.0 license applied
- Finalize commandline interface
- Sanitizing some setup.py and test suite running
- Added Travis CI support


test_csv2es.py
## Copyright 2015 Ray Holder
##
## Licensed under the Apache License, Version 2.0 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
## http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.

import csv2es
import unittest


class TestDelimiter(unittest.TestCase):

    def test_sanitize(self):
        self.assertEqual(None, csv2es.sanitize_delimiter(None, False))
        self.assertEqual(str('\t'), csv2es.sanitize_delimiter(None, True))
        self.assertEqual(str('|'), csv2es.sanitize_delimiter('|', False))
        self.assertEqual(str('|'), csv2es.sanitize_delimiter(u'|', False))
        self.assertEqual(str('\t'), csv2es.sanitize_delimiter('|', True))
        self.assertEqual(str('\t'), csv2es.sanitize_delimiter('||', True))
        self.assertRaises(Exception, csv2es.sanitize_delimiter, '||', False)


class TestLoading(unittest.TestCase):

    def test_csv(self):
        # TODO fill this in
        self.assertTrue(True)

    def test_tsv(self):
        # TODO fill this in
        self.assertTrue(True)


if __name__ == '__main__':
    unittest.main()


json2es.py
## Copyright 2015 Ray Holder
##
## Licensed under the Apache License, Version 2.0 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
## http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.

import csv
import json
import sys
from threading import local

import click
from joblib import Parallel, delayed
from pyelasticsearch import ElasticSearch
from pyelasticsearch import bulk_chunks
from pyelasticsearch import ElasticHttpNotFoundError
from pyelasticsearch import IndexAlreadyExistsError
from retrying import retry


__version__ = '1.0.1'
thread_local = local()


def echo(message, quiet):
    """
    Print the given message to standard out via click unless quiet is True.

    :param message: the message to print out to the console
    :param quiet: don't print the message when this is True
    """
    if not quiet:
        click.echo(message)


def documents_from_file(es, filename, quiet):
    """
    Return a generator for pulling rows from a given delimited file.

    :param es: an ElasticSearch client
    :param filename: the name of the file to read from or '-' if stdin
    :param delimiter: the delimiter to use
    :param quiet: don't output anything to the console when this is True
    :return: generator returning document-indexing operations
    """
    def all_docs():
        with open(filename, 'rb') if filename != '-' else sys.stdin as doc_file:

            count = 0
            docs = json.load(doc_file)
            for doc in docs:
                count += 1
                if count % 10000 == 0:
                    echo('Sent documents: ' + str(count), quiet)
                yield es.index_op(doc)

    return all_docs


@retry(wait_exponential_multiplier=1000, wait_exponential_max=10000, stop_max_attempt_number=10)
def local_bulk(host, index_name, doc_type, chunk):
    """
    Bulk upload the given chunk, creating a thread local ElasticSearch instance
    for the target host if one does not already exist. Retry this function at
    least 10 times with exponential backoff.

    :param host: the target Elasticsearch host
    :param index_name: the index name
    :param doc_type: the document type
    :param chunk: the chunk of documents to bulk upload
    """
    if not hasattr(thread_local, 'es'):
        thread_local.es = ElasticSearch(host)

    thread_local.es.bulk(chunk, index=index_name, doc_type=doc_type)


def perform_bulk_index(host, index_name, doc_type, doc_fetch, docs_per_chunk, bytes_per_chunk, parallel):
    """
    Chunk up documents and send them to Elasticsearch in bulk.

    :param host: the target Elasticsearch host
    :param index_name: the target index name
    :param doc_type: the target document type
    :param doc_fetch: a function to call to fetch documents
    :param docs_per_chunk: the number of documents per chunk to upload
    :param bytes_per_chunk: the max bytes per chunk to upload
    :param parallel: the number of bulk uploads to do at the same time
    """
    Parallel(n_jobs=parallel)(
        delayed(local_bulk)(host, index_name, doc_type, chunk)
        for chunk in bulk_chunks(doc_fetch(),
                                 docs_per_chunk=docs_per_chunk,
                                 bytes_per_chunk=bytes_per_chunk))



@click.command()
@click.option('--index-name', required=True,
              help='Index name to load data into         ')
@click.option('--doc-type', required=True,
              help='The document type (like user_records)')
@click.option('--import-file', required=True,
              help='File to import (or \'-\' for stdin)    ')
@click.option('--mapping-file', required=False,
              help='JSON mapping file for index')
@click.option('--host', default='http://127.0.0.1:9200/', required=False,
              help='The Elasticsearch host (http://127.0.0.1:9200/)')
@click.option('--docs-per-chunk', default=5000, required=False,
              help='The documents per chunk to upload (5000)')
@click.option('--bytes-per-chunk', default=100000, required=False,
              help='The bytes per chunk to upload (100000)')
@click.option('--parallel', default=1, required=False,
              help='Parallel uploads to send at once, defaults to 1')
@click.option('--delete-index', is_flag=True, required=False,
              help='Delete existing index if it exists')
@click.option('--quiet', is_flag=True, required=False,
              help='Minimize console output')
@click.version_option(version=__version__, )
def cli(index_name, delete_index, mapping_file, doc_type, import_file,
        host, docs_per_chunk, bytes_per_chunk, parallel, quiet):
    """
    Bulk import a delimited file into a target Elasticsearch instance. Common
    delimited files include things like CSV and TSV.

    \b
    Load a CSV file:
      csv2es --index-name potatoes --doc-type potato --import-file potatoes.csv
    \b
    For a TSV file, note the tab delimiter option
      csv2es --index-name tomatoes --doc-type tomato --import-file tomatoes.tsv --tab
    \b
    For a nifty pipe-delimited file (delimiters must be one character):
      csv2es --index-name pipes --doc-type pipe --import-file pipes.psv --delimiter '|'

    """

    echo('Using host: ' + host, quiet)
    es = ElasticSearch(host)

    if delete_index:
        try:
            es.delete_index(index_name)
            echo('Deleted: ' + index_name, quiet)
        except ElasticHttpNotFoundError:
            echo('Index ' + index_name + ' not found, nothing to delete', quiet)

    try:
        es.create_index(index_name)
        echo('Created new index: ' + index_name, quiet)
    except IndexAlreadyExistsError:
        echo('Index ' + index_name + ' already exists', quiet)

    echo('Using document type: ' + doc_type, quiet)
    if mapping_file:
        echo('Applying mapping from: ' + mapping_file, quiet)
        with open(mapping_file) as f:
            mapping = json.loads(f.read())
        es.put_mapping(index_name, doc_type, mapping)

    documents = documents_from_file(es, import_file, quiet)
    perform_bulk_index(host, index_name, doc_type, documents, docs_per_chunk, bytes_per_chunk, parallel)


if __name__ == "__main__":
    cli()


csv2es.py
## Copyright 2015 Ray Holder
##
## Licensed under the Apache License, Version 2.0 (the "License");
## you may not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
## http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS,
## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.

import csv
import json
import sys
from threading import local

import click
from joblib import Parallel, delayed
from pyelasticsearch import ElasticSearch
from pyelasticsearch import bulk_chunks
from pyelasticsearch import ElasticHttpNotFoundError
from pyelasticsearch import IndexAlreadyExistsError
from retrying import retry


__version__ = '1.0.2'
thread_local = local()


def echo(message, quiet):
    """
    Print the given message to standard out via click unless quiet is True.

    :param message: the message to print out to the console
    :param quiet: don't print the message when this is True
    """
    if not quiet:
        click.echo(message)


def documents_from_file(es, filename, delimiter, quiet):
    """
    Return a generator for pulling rows from a given delimited file.

    :param es: an ElasticSearch client
    :param filename: the name of the file to read from or '-' if stdin
    :param delimiter: the delimiter to use
    :param quiet: don't output anything to the console when this is True
    :return: generator returning document-indexing operations
    """
    def all_docs():
        with open(filename, 'rb') if filename != '-' else sys.stdin as doc:
            # delimited file should include the field names as the first row
            fieldnames = doc.next().strip().split(delimiter)
            echo('Using the following ' + str(len(fieldnames)) + ' fields:',
                 quiet)
            for fieldname in fieldnames:
                echo(fieldname, quiet)

            reader = csv.DictReader(doc, delimiter=delimiter,
                                    fieldnames=fieldnames)
            count = 0
            for row in reader:
                count += 1
                if count % 10000 == 0:
                    echo('Sent documents: ' + str(count), quiet)
                yield es.index_op(row)

    return all_docs


@retry(wait_exponential_multiplier=1000, wait_exponential_max=10000,
       stop_max_attempt_number=10)
def local_bulk(host, index_name, doc_type, chunk):
    """
    Bulk upload the given chunk, creating a thread local ElasticSearch instance
    for the target host if one does not already exist. Retry this function at
    least 10 times with exponential backoff.

    :param host: the target Elasticsearch host
    :param index_name: the index name
    :param doc_type: the document type
    :param chunk: the chunk of documents to bulk upload
    """
    if not hasattr(thread_local, 'es'):
        thread_local.es = ElasticSearch(host)

    thread_local.es.bulk(chunk, index=index_name, doc_type=doc_type)


def perform_bulk_index(host, index_name, doc_type, doc_fetch, docs_per_chunk,
                       bytes_per_chunk, parallel):
    """
    Chunk up documents and send them to Elasticsearch in bulk.

    :param host: the target Elasticsearch host
    :param index_name: the target index name
    :param doc_type: the target document type
    :param doc_fetch: a function to call to fetch documents
    :param docs_per_chunk: the number of documents per chunk to upload
    :param bytes_per_chunk: the max bytes per chunk to upload
    :param parallel: the number of bulk uploads to do at the same time
    """
    Parallel(n_jobs=parallel)(
        delayed(local_bulk)(host, index_name, doc_type, chunk)
        for chunk in bulk_chunks(doc_fetch(),
                                 docs_per_chunk=docs_per_chunk,
                                 bytes_per_chunk=bytes_per_chunk))


def sanitize_delimiter(delimiter, is_tab):
    """
    Return a single character delimiter from the given (possibly unicode)
    string. If is_tab is True, always return a single tab. If delimiter is None
    then return None. Raise an Exception if the delimiter can't be converted to
    a single character.

    Why is this so complicated with some kind of special artisan tab handling?
    Well, passing in a tab character as a delimiter from the commandline as
    exactly 1 character, unescaped, and sanitized of unicode wonkery was
    non-trivial for me to remember. I didn't want to have to ask people to
    "obviously" just pass in --delimiter $'\t' for the extremely common case of
    wanting to load a TSV file. I'm sure this could be better.

    :param delimiter: the delimiter to use
    :param is_tab: if this is True, just return a tab character
    :return a one character delimiter
    """

    if is_tab:
        return str('\t')

    if delimiter is None:
        return None
    else:
        d = str(delimiter)
        if len(d) == 1:
            return d
        else:
            raise Exception('Delimiter cannot be more than 1 character.')


@click.command()
@click.option('--index-name', required=True,
              help='Index name to load data into         ')
@click.option('--doc-type', required=True,
              help='The document type (like user_records)')
@click.option('--import-file', required=True,
              help='File to import (or \'-\' for stdin)    ')
@click.option('--mapping-file', required=False,
              help='JSON mapping file for index')
@click.option('--settings-file', required=False,
              help='JSON settings file for index')
@click.option('--delimiter', required=False,
              help='The field delimiter to use, defaults to CSV')
@click.option('--tab', is_flag=True, required=False,
              help='Assume tab-separated, overrides delimiter')
@click.option('--host', default='http://127.0.0.1:9200/', required=False,
              help='The Elasticsearch host (http://127.0.0.1:9200/)')
@click.option('--docs-per-chunk', default=5000, required=False,
              help='The documents per chunk to upload (5000)')
@click.option('--bytes-per-chunk', default=100000, required=False,
              help='The bytes per chunk to upload (100000)')
@click.option('--parallel', default=1, required=False,
              help='Parallel uploads to send at once, defaults to 1')
@click.option('--delete-index', is_flag=True, required=False,
              help='Delete existing index if it exists')
@click.option('--quiet', is_flag=True, required=False,
              help='Minimize console output')
@click.version_option(version=__version__, )
def cli(index_name, delete_index, mapping_file, settings_file, doc_type,
        import_file, delimiter, tab, host, docs_per_chunk, bytes_per_chunk,
        parallel, quiet):
    """
    Bulk import a delimited file into a target Elasticsearch instance. Common
    delimited files include things like CSV and TSV.

    \b
    Load a CSV file:
      csv2es --index-name potatoes --doc-type potato --import-file potatoes.csv
    \b
    For a TSV file, note the tab delimiter option
      csv2es --index-name tomatoes --doc-type tomato \
             --import-file tomatoes.tsv --tab
    \b
    For a nifty pipe-delimited file (delimiters must be one character):
      csv2es --index-name pipes --doc-type pipe --import-file pipes.psv \
             --delimiter '|'

    """

    echo('Using host: ' + host, quiet)
    es = ElasticSearch(host)

    if delete_index:
        try:
            es.delete_index(index_name)
            echo('Deleted: ' + index_name, quiet)
        except ElasticHttpNotFoundError:
            echo('Index ' + index_name + ' not found, nothing to delete',
                 quiet)

    try:
        if settings_file:
            echo('Applying mapping from: ' + settings_file, quiet)
            with open(settings_file) as f:
                settings = json.loads(f.read())
            es.create_index(index_name, settings)
        else:
            es.create_index(index_name)
        echo('Created new index: ' + index_name, quiet)
    except IndexAlreadyExistsError:
        echo('Index ' + index_name + ' already exists', quiet)

    echo('Using document type: ' + doc_type, quiet)
    if mapping_file:
        echo('Applying mapping from: ' + mapping_file, quiet)
        with open(mapping_file) as f:
            mapping = json.loads(f.read())
        es.put_mapping(index_name, doc_type, mapping)

    target_delimiter = sanitize_delimiter(delimiter, tab)
    documents = documents_from_file(es, import_file, target_delimiter, quiet)
    perform_bulk_index(host, index_name, doc_type, documents, docs_per_chunk,
                       bytes_per_chunk, parallel)


if __name__ == "__main__":
    cli()
